host/logmux/logmux.go

```diff
@@ -0,0 +1,138 @@
+package logmux
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/flynn/flynn/discoverd/client"
+	"github.com/flynn/flynn/pkg/syslog"
+	"github.com/flynn/flynn/pkg/syslog/rfc5424"
+)
+
+type LogMux struct {
+	logc, bufc chan *rfc5424.Message
+
+	doneo sync.Once
+	donec chan struct{}
+}
+
+func New(discd *discoverd.Client, bufferSize int) (*LogMux, error) {
+	slc, err := dialService(discd)
+	if err != nil {
+		return nil, err
+	}
+
+	mux := &LogMux{
+		logc:  make(chan *rfc5424.Message),
+		bufc:  make(chan *rfc5424.Message, bufferSize),
+		donec: make(chan struct{}),
+	}
+
+	go mux.drainLogc(discd, slc)
+	go mux.drainBufc()
+
+	return mux, nil
+}
+
+func (m *LogMux) drainLogc(discd *discoverd.Client, slc *syslog.Client) {
+	eventc := make(chan *discoverd.Event)
+
+	var err error
+	for {
+		select {
+		case msg := <-m.logc:
+			if msg == nil {
+				if slc.Close(); err != nil {
+					panic(err)
+				}
+				close(m.donec)
+				return
+			}
+
+			if _, err = slc.WriteMessage(msg); err != nil {
+				panic(err)
+			}
+		case event := <-eventc:
+			if event.Kind == discoverd.EventKindLeader {
+				if slc.Close(); err != nil {
+					panic(err)
+				}
+				if slc, err = dialService(discd); err != nil {
+					panic(err)
+				}
+			}
+		}
+	}
+}
+
+func dialService(discd *discoverd.Client) (*syslog.Client, error) {
+	srv := discd.Service("logaggregator")
+	ldr, err := srv.Leader()
+	if err != nil {
+		return nil, err
+	}
+
+	return syslog.Dial(ldr.Addr, nil)
+}
+
+func (m *LogMux) drainBufc() {
+	for {
+		msg := <-m.bufc
+		if msg == nil {
+			close(m.logc)
+			return
+		}
+
+		select {
+		case m.logc <- msg:
+		default:
+			select {
+			case m.logc <- msg:
+			case <-m.logc:
+				m.logc <- msg
+			}
```

Comment:
Something is off here. Why is the inner select also selecting on the same thing as the outer select?



AFAICT this code is here to make sure that you're reading off the FD as fast as possible, throwing away log lines if you can't process them fast enough.

It seems like a better place to do that is in `follow()`, where you can do the following to just throw away log lines if `bufc` doesn't have space for them:

```go
select {
case m.bufc <- msg:
default:
	// throw away msg
}
```

Then maybe you don't even need 2 separate channels (`bufc` and `logc`).